controller/deployments.go

```diff
@@ -404,6 +409,138 @@ func (c *DeploymentsController) ShowAllEnvironments(ctx *app.ShowAllEnvironments
+			return
+		}
+
+		store, stopWs := kc.WatchEventsInNamespace(ctx.EnvName)
+		defer close(stopWs)
+
+		go func() {
+			for {
+				var m string
+				err := websocket.Message.Receive(ws, &m)
+				if err != nil {
+					if err != io.EOF {
+						log.Error(ctx, map[string]interface{}{
+							"err": err,
+						}, "error reading from websocket")
+					}
+					store.Close()
+					return
+				}
+			}
+		}()
+
+		for {
+			item, err := store.Pop(cache.PopProcessFunc(func(item interface{}) error {
+				return nil
+			}))
+			if err != nil {
+				if err != cache.FIFOClosedError {
+					log.Error(ctx, map[string]interface{}{
+						"err": err,
+					}, "error receiving events")
+
+					sendWebsocketJSON(ctx, ws, map[string]interface{}{"error": "unable to access Kubernetes events"})
+				}
+				return
+			}
+
+			eventItem, err := transformItem(item)
+			if err != nil {
+				sendWebsocketJSON(ctx, ws, map[string]interface{}{"error": "unable to parse Kubernetes event"})
+			} else {
+				err = websocket.JSON.Send(ws, eventItem)
+				if err != nil {
+					log.Error(ctx, map[string]interface{}{
+						"err": err,
+					}, "error sending events")
+					return
+				}
+			}
+		}
+	}
+}
+
+//DeploymentsEvent is the transformed Kubernetes v1.Event item
+type DeploymentsEvent struct {
+	// The object that this event is about.
+	InvolvedObject v1.ObjectReference `json:"involvedObject" protobuf:"bytes,2,opt,name=involvedObject"`
+
+	// This should be a short, machine understandable string that gives the reason
+	// for the transition into the object's current status.
+	// TODO: provide exact specification for format.
+	// +optional
+	Reason string `json:"reason,omitempty" protobuf:"bytes,3,opt,name=reason"`
+
+	// A human-readable description of the status of this operation.
+	// TODO: decide on maximum length.
+	// +optional
+	Message string `json:"message,omitempty" protobuf:"bytes,4,opt,name=message"`
+
+	// The number of times this event has occurred.
+	// +optional
+	Count int32 `json:"count,omitempty" protobuf:"varint,8,opt,name=count"`
+
+	// Type of this event (Normal, Warning), new types could be added in the future
+	// +optional
+	Type string `json:"type,omitempty" protobuf:"bytes,9,opt,name=type"`
+
+	// CreationTimestamp is a timestamp representing the server time when this object was
+	// created. It is not guaranteed to be set in happens-before order across separate operations.
+	// Clients may not set this value. It is represented in RFC3339 form and is in UTC.
+	//
+	// Populated by the system.
+	// Read-only.
+	// Null for lists.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	CreationTimestamp metaV1.Time `json:"creationTimestamp,omitempty" protobuf:"bytes,8,opt,name=creationTimestamp"`
+}
+
+func transformItem(item interface{}) (*DeploymentsEvent, error) {
+	if event, ok := item.(*v1.Event); ok {
```

Comment:
Item is guaranteed to be a `v1.Event` isn't it? We could get rid of the type-assertion and error return value here.

Reply:
I'm not sure the best way to go about this. The input for this comes from the Kubernetes Store that is returned from
```
store, stopWs := kc.WatchEventsInNamespace(ctx.EnvName)
```

The definition for `FIFO` `pop` is `func (f *FIFO) Pop(process PopProcessFunc) (interface{}, error) {`

Is there a way I can rework it to give me a `*v1.Event` so I don't need to to the type transformation? Actually, I guess I can do it earlier so `transformItem` takes a `*v1.Event`. That will be less confusing.

Reply:
Ah, sorry this comment was meant for the `transformItem` in the unit tests, where the argument passed is of Event type already.

Reply:
Oh, no worries. I've changed the one in the unit tests. I've also reworked the one in the controller too so the type transform is earlier.